azurerm/resource_arm_ad_application.go

```diff
@@ -0,0 +1,325 @@
+package azurerm
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/Azure/azure-sdk-for-go/services/graphrbac/1.6/graphrbac"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmAdApplication() *schema.Resource {
+	return &schema.Resource{
+		Create:        resourceArmAdApplicationCreate,
+		Read:          resourceArmAdApplicationRead,
+		Update:        resourceArmAdApplicationUpdate,
+		Delete:        resourceArmAdApplicationDelete,
+		CustomizeDiff: customizeDiffAd,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"display_name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"homepage": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"identifier_uris": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				MinItems: 1,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+
+			"reply_urls": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				MinItems: 1,
```

Comment:
given this has to be specified at create time - isn't this Required?

Reply:
You can edit these at any time to add/remove as needed, and they're not required for creation. I believe the bare minimum properties required for creating an application are the display name, 1 identifier URI, and the homepage. Those last two properties attempt to set default values for you based on the display name you give in their expand functions as a convenience, but the expand for reply_urls doesn't as it's not required. I modeled the default functionality based on how Azure CLI's [az ad app create](https://docs.microsoft.com/en-us/cli/azure/ad/app?view=azure-cli-latest) works.